### DISCUSSION TITLE
Weekly #4

### DISCUSSION THREAD

**Discussion by Ninetilt** [Ninetilt/Project-TDC#6] [General]
### **Übersicht**
In diesem Blog-Post zeigen wir euch unser Klassendiagramm präsentieren, darauf eingehen welche Prinzipien wir bei der Erstellung verwendet haben und ein Paar Worte über das Anlegen unseres Projektes verlieren.

### **Klassendiagramm** ([UML](https://github.com/Ninetilt/Project-TDC/blob/main/docs/uml/PDF/TDC-Class.pdf))
Für die Erstellung unserer Klassen haben wir anhand unseres [Mockups ](https://github.com/Ninetilt/Project-TDC/tree/main/docs/ui)gearbeitet, um uns einen Überblick darüber zu verschaffen was wir alles benötigen. Die Klassen haben wir bereits mit allen (vom aktuellen Stand) benötigten Variablen und Datenstrukturen ausgestattet. Setter und Getter Methoden sind auch bereits alle mit einbegriffen. Darüber hinaus haben wir auch einige klassenspezifische Methoden deklariert, die nach dem aktuellen Umfang für die geplanten Funktionen ausreichen sollten. Falls und beim Programmieren auffallen sollte, dass dies nicht der Fall ist, wird das Klassendiagramm im nachhinein noch angepasst. Bei den Relationen zwischen den Klassen arbeiten wir hauptsächlich mit Aggregationen, aber auch mit einer Komposition (zw. ToDoList und ListItem) und einer Generalisierung/Vererbung (Account erbt von Profil).

Unsere Sequenzdiagramme haben wir entsprechend angepasst. Diese wurden dadurch auch etwas komplexer. Die neuen Diagramme findet ihr hier: [AddFriend ](https://github.com/Ninetilt/Project-TDC/blob/main/docs/uml/PDF/AddFriendSequence.pdf) | [FriendList](https://github.com/Ninetilt/Project-TDC/blob/main/docs/uml/PDF/FinishListSequence.pdf)
Zum Vergleich lassen wir die alten Version zunächst auch online: [AddFriend ](https://github.com/Ninetilt/Project-TDC/blob/main/docs/uml/PDF/AddFriendSequenceDeprecated.pdf) | [FriendList](https://github.com/Ninetilt/Project-TDC/blob/main/docs/uml/PDF/FinishListSequenceDeprecated.pdf)

Die **SOLID-Prinzipien** nach denen wir uns entschieden habe zu arbeiten sind das **S** und das **O**. Nach dem "Single Responsibility" Prinzip hat jede Klasse nur eine Verantwrotung und nach dem "Open-closed" Prinzip sind Klassen erweiterbar, aber nicht modifizierbar.

![TDC-Class](https://github.com/user-attachments/assets/e1082d47-dd02-452a-9ced-2874c8d8e59a)

### **Aufsetzen des Projekts**
Da wir nun mit dem Erstellen der Klassen auch mit dem Porgrammien begonnen haben, haben wir direkt das Projekt mit Hilfe von Visual Studio und Git in C# aufgesetzt, um auch zukünftige Programmierarbeit nahtlos einzubinden. Ein großer Aufwand war das nicht, aber dennoch Wert es hier im Blog einmal zu erwähnen.

Falls ihr noch Fragen zu unserem Klassendiagramm oder zum Projekt generell habt, schreibt es gerne einfach in die Kommentare.

**Comment by j1tbk** [comment]
Das Klassendiagramm zeigt eine sehr durchdachte und sorgfältig strukturierte Designlösung, die eine Vielzahl von Klassen und deren Beziehungen zueinander klar und präzise darstellt. Die Organisation der Klassen und die logischen Verbindungen zwischen ihnen spiegeln ein komplexes, aber auch leicht nachvollziehbares System wider, das verschiedene Aspekte eines Accounts- und Profilverwaltungssystems abdeckt.

Besonders beeindruckend ist die Aufteilung der Verantwortlichkeiten: Jede Klasse hat klar definierte Aufgaben und Methoden, die eng an die spezifische Funktion im Gesamtsystem gekoppelt sind. Zum Beispiel ist die Klasse ProfileRepository ausschließlich für die Verwaltung der Profile zuständig, während AccountRepository die Verwaltung der Accounts übernimmt. Diese Trennung erleichtert die Wartung und Erweiterung des Systems und fördert die Wiederverwendbarkeit der Komponenten.

Liebe Grüße
Project Health

**Comment by DonKir15** [comment]
Der Blogpost zeigt ein gut durchdachtes und strukturiertes Vorgehen beim Aufbau des Projekts. Es ist positiv, dass ihr sorgfältig anhand von Mockups arbeitet, um alle nötigen Klassen, Variablen und Datenstrukturen zu identifizieren.  Die Planung, das Klassendiagramm bei Bedarf anzupassen, zeugt von Flexibilität und einem guten Umgang mit möglichen Anpassungen im Entwicklungsprozess.
Besonders gut ist die Anwendung der SOLID-Prinzipien, insbesondere des "Single Responsibility"- und des "Open-Closed"-Prinzips. Diese Designentscheidungen helfen dabei, die Codequalität zu erhöhen und die Erweiterbarkeit des Projekts sicherzustellen.

**Comment by TimHitschfeld** [comment]
Das Klassendiagramm und die Anwendung der SOLID-Prinzipien haben eine klare und gut strukturierte Architektur hervorgebracht, die besonders die Trennung von Verantwortlichkeiten betont. Dies fördert nicht nur die Wartbarkeit des Codes, sondern auch seine Erweiterbarkeit, was für die Skalierbarkeit des Projekts wichtig ist. Die Integration der Mockups und die bewusste Entscheidung, Raum für zukünftige Anpassungen zu lassen, zeigen eine vorausschauende Denkweise. Es ist klar, dass ihr euch auf nachhaltigen und flexiblen Code konzentriert, der zukünftige Anforderungen gut abdecken kann. Weiter so, ich freue mich auf die nächsten Schritte!

Viele Grüße,
Euer TripAhead Team /TH
